File: js/moq/src/ietf/subscriber.ts
import { type AnnouncedConsumer, AnnouncedProducer } from "../announced.ts";
import { type BroadcastConsumer, BroadcastProducer } from "../broadcast.ts";
import { GroupProducer } from "../group.ts";
import * as Path from "../path.ts";
import type { Reader } from "../stream.ts";
import type { TrackProducer } from "../track.ts";
import { error } from "../util/error.ts";
import type { Announce, Unannounce } from "./announce.ts";
import type * as Control from "./control.ts";
import { Frame, type Group } from "./object.ts";
import { Subscribe, type SubscribeDone, type SubscribeError, type SubscribeOk, Unsubscribe } from "./subscribe.ts";
import type { SubscribeAnnouncesError, SubscribeAnnouncesOk } from "./subscribe_announces.ts";
import type { TrackStatus } from "./track.ts";

/**
 * Handles subscribing to broadcasts using moq-transport protocol with lite-compatibility restrictions.
 *
 * @internal
 */
export class Subscriber {
	#control: Control.Stream;
	#root: Path.Valid;

	// Our subscribed tracks - keyed by subscription ID
	#subscribes = new Map<bigint, TrackProducer>();
	#subscribeNext = 0n;

	// Track subscription responses - keyed by subscription ID
	#subscribeCallbacks = new Map<
		bigint,
		{
			resolve: (msg: SubscribeOk) => void;
			reject: (msg: Error) => void;
		}
	>();

	#announced: AnnouncedProducer;

	/**
	 * Creates a new Subscriber instance.
	 * @param quic - The WebTransport session to use
	 * @param control - The control stream writer for sending control messages
	 *
	 * @internal
	 */
	constructor(control: Control.Stream, root: Path.Valid) {
		this.#control = control;
		this.#root = root;
		this.#announced = new AnnouncedProducer();
		//void this.#runAnnounced();
	}

	/* TODO once the remote server actually supports it
	async #runAnnounced() {
		// Send me everything at the root.
		const msg = new SubscribeAnnounces(this.#root);
		await Control.write(this.#control, msg);
	}
	*/

	/**
	 * Gets an announced reader for the specified prefix.
	 * @param prefix - The prefix for announcements
	 * @returns An AnnounceConsumer instance
	 */
	announced(prefix: Path.Valid = Path.empty()): AnnouncedConsumer {
		const full = Path.join(this.#root, prefix);
		return this.#announced.consume(full);
	}

	/**
	 * Consumes a broadcast from the connection.
	 *
	 * NOTE: This is not automatically deduplicated.
	 * If to consume the same broadcast twice, and subscribe to the same tracks twice, then network usage is doubled.
	 * However, you can call `clone()` on the consumer to deduplicate and share the same handle.
	 *
	 * @param name - The name of the broadcast to consume
	 * @returns A BroadcastConsumer instance
	 */
	consume(broadcast: Path.Valid): BroadcastConsumer {
		const producer = new BroadcastProducer();
		const consumer = producer.consume();

		producer.unknownTrack((track) => {
			// Save the track in the cache to deduplicate.
			// NOTE: We don't clone it (yet) so it doesn't count as an active consumer.
			// When we do clone it, we'll only get the most recent (consumed) group.
			producer.insertTrack(track.consume());

			// Perform the subscription in the background.
			this.#runSubscribe(broadcast, track).finally(() => {
				try {
					producer.removeTrack(track.name);
				} catch {
					// Already closed.
					console.warn("track already removed");
				}
			});
		});

		// Close when the producer has no more consumers.
		producer.unused().finally(() => {
			producer.close();
		});

		return consumer;
	}

	async #runSubscribe(broadcast: Path.Valid, track: TrackProducer) {
		const subscribeId = this.#subscribeNext++;

		// Save the writer so we can append groups to it.
		this.#subscribes.set(subscribeId, track);

		const msg = new Subscribe(subscribeId, subscribeId, broadcast, track.name, track.priority);

		// Send SUBSCRIBE message on control stream and wait for response
		const responsePromise = new Promise<SubscribeOk>((resolve, reject) => {
			this.#subscribeCallbacks.set(subscribeId, { resolve, reject });
		});

		await this.#control.write(msg);

		try {
			await responsePromise;
			await track.unused();

			track.close();

			const msg = new Unsubscribe(subscribeId);
			await this.#control.write(msg);
		} catch (err) {
			const e = error(err);
			track.abort(e);
		} finally {
			this.#subscribes.delete(subscribeId);
			this.#subscribeCallbacks.delete(subscribeId);
		}
	}

	/**
	 * Handles a SUBSCRIBE_OK control message received on the control stream.
	 * @param msg - The SUBSCRIBE_OK message
	 *
	 * @internal
	 */
	async handleSubscribeOk(msg: SubscribeOk) {
		const callback = this.#subscribeCallbacks.get(msg.subscribeId);
		if (callback) {
			callback.resolve(msg);
		}
	}

	/**
	 * Handles a SUBSCRIBE_ERROR control message received on the control stream.
	 * @param msg - The SUBSCRIBE_ERROR message
	 *
	 * @internal
	 */
	async handleSubscribeError(msg: SubscribeError) {
		const callback = this.#subscribeCallbacks.get(msg.subscribeId);
		if (callback) {
			callback.reject(new Error(`SUBSCRIBE_ERROR: code=${msg.errorCode} reason=${msg.reasonPhrase}`));
		}
	}

	/**
	 * Handles an ObjectStream message (moq-transport equivalent of moq-lite Group).
	 * @param msg - The ObjectStream message
	 * @param stream - The stream to read object data from
	 *
	 * @internal
	 */
	async handleGroup(group: Group, stream: Reader) {
		const producer = new GroupProducer(group.groupId);

		try {
			const track = this.#subscribes.get(group.trackAlias);
			if (!track) {
				throw new Error(`unknown track: alias=${group.trackAlias}`);
			}

			// Convert to Group (moq-lite equivalent)
			track.insertGroup(producer.consume());

			// Read objects from the stream until end of group
			for (;;) {
				const done = await Promise.race([stream.done(), track.unused(), producer.unused()]);
				if (done !== false) break;

				const frame = await Frame.decode(stream);
				if (frame.payload === undefined) break;

				// Treat each object payload as a frame
				producer.writeFrame(frame.payload);
			}

			producer.close();
		} catch (err: unknown) {
			const e = error(err);
			producer.abort(e);
			stream.stop(e);
		}
	}

	/**
	 * Handles a SUBSCRIBE_DONE control message received on the control stream.
	 * @param msg - The SUBSCRIBE_DONE message
	 */
	async handleSubscribeDone(msg: SubscribeDone) {
		// For lite compatibility, we treat this as subscription completion
		const callback = this.#subscribeCallbacks.get(msg.subscribeId);
		if (callback) {
			callback.reject(new Error(`SUBSCRIBE_DONE: code=${msg.statusCode} reason=${msg.reasonPhrase}`));
		}
	}

	/**
	 * Handles an ANNOUNCE control message received on the control stream.
	 * @param msg - The ANNOUNCE message
	 */
	async handleAnnounce(msg: Announce) {
		this.#announced.write({
			name: msg.trackNamespace,
			active: true,
		});
	}

	/**
	 * Handles an UNANNOUNCE control message received on the control stream.
	 * @param msg - The UNANNOUNCE message
	 */
	async handleUnannounce(msg: Unannounce) {
		this.#announced.write({
			name: msg.trackNamespace,
			active: false,
		});
	}

	async handleSubscribeAnnouncesOk(_msg: SubscribeAnnouncesOk) {
		// TODO
	}

	async handleSubscribeAnnouncesError(_msg: SubscribeAnnouncesError) {
		// TODO
	}

	/**
	 * Handles a TRACK_STATUS control message received on the control stream.
	 * @param msg - The TRACK_STATUS message
	 */
	async handleTrackStatus(_msg: TrackStatus) {
		// TODO
	}
}

File: rs/moq/src/ietf/subscriber.rs
use std::{
	collections::{hash_map::Entry, HashMap},
	sync::{atomic, Arc},
};

use crate::{
	coding::Reader,
	ietf::{self, Control},
	model::BroadcastProducer,
	Broadcast, Error, Frame, FrameProducer, Group, GroupProducer, OriginProducer, Path, PathOwned, TrackProducer,
};

use web_async::Lock;

#[derive(Clone)]
pub(super) struct Subscriber<S: web_transport_trait::Session> {
	session: S,

	origin: Option<OriginProducer>,
	subscribes: Lock<HashMap<u64, TrackProducer>>,
	next_id: Arc<atomic::AtomicU64>,

	producers: Lock<HashMap<PathOwned, BroadcastProducer>>,
	control: Control,
}

impl<S: web_transport_trait::Session> Subscriber<S> {
	pub fn new(session: S, origin: Option<OriginProducer>, control: Control) -> Self {
		Self {
			session,
			origin,
			subscribes: Default::default(),
			next_id: Default::default(),
			producers: Default::default(),
			control,
		}
	}

	pub fn recv_announce(&mut self, msg: ietf::Announce) -> Result<(), Error> {
		let origin = match &self.origin {
			Some(origin) => origin,
			None => {
				self.control.send(
					ietf::MessageId::AnnounceError,
					ietf::AnnounceError {
						track_namespace: msg.track_namespace,
						error_code: 404,
						reason_phrase: "Publish only".into(),
					},
				)?;

				return Ok(());
			}
		};

		let path = msg.track_namespace.to_owned();
		tracing::debug!(broadcast = %origin.absolute(&path), suffix = %path, "announce");

		let broadcast = Broadcast::produce();

		// Make sure the peer doesn't double announce.
		match self.producers.lock().entry(path.to_owned()) {
			Entry::Occupied(_) => return Err(Error::Duplicate),
			Entry::Vacant(entry) => entry.insert(broadcast.producer.clone()),
		};

		// Run the broadcast in the background until all consumers are dropped.
		origin.publish_broadcast(path.clone(), broadcast.consumer);

		self.control.send(
			ietf::MessageId::AnnounceOk,
			ietf::AnnounceOk {
				track_namespace: path.clone(),
			},
		)?;

		web_async::spawn(self.clone().run_broadcast(path, broadcast.producer));

		Ok(())
	}

	pub fn recv_unannounce(&mut self, msg: ietf::Unannounce) -> Result<(), Error> {
		let origin = match &self.origin {
			Some(origin) => origin,
			None => return Ok(()),
		};

		let path = msg.track_namespace.to_owned();
		tracing::debug!(broadcast = %origin.absolute(&path), "unannounced");

		// Close the producer.
		let mut producer = self.producers.lock().remove(&path).ok_or(Error::NotFound)?;

		producer.close();

		Ok(())
	}

	pub fn recv_subscribe_ok(&mut self, _msg: ietf::SubscribeOk) -> Result<(), Error> {
		// Don't care.
		Ok(())
	}

	pub fn recv_subscribe_error(&mut self, msg: ietf::SubscribeError<'_>) -> Result<(), Error> {
		if let Some(track) = self.subscribes.lock().remove(&msg.subscribe_id) {
			track.abort(Error::Cancel);
		}

		Ok(())
	}

	pub fn recv_subscribe_done(&mut self, msg: ietf::SubscribeDone<'_>) -> Result<(), Error> {
		if let Some(track) = self.subscribes.lock().remove(&msg.subscribe_id) {
			track.close();
		}

		Ok(())
	}

	pub async fn run(self) -> Result<(), Error> {
		loop {
			let stream = self
				.session
				.accept_uni()
				.await
				.map_err(|err| Error::Transport(Arc::new(err)))?;

			let stream = Reader::new(stream);
			let this = self.clone();

			web_async::spawn(async move {
				if let Err(err) = this.run_uni_stream(stream).await {
					tracing::debug!(%err, "error running uni stream");
				}
			});
		}
	}

	async fn run_uni_stream(mut self, mut stream: Reader<S::RecvStream>) -> Result<(), Error> {
		let kind = stream.decode().await?;

		let res = match kind {
			ietf::Group::STREAM_TYPE => self.recv_group(&mut stream).await,
			_ => return Err(Error::UnexpectedStream),
		};

		if let Err(err) = res {
			stream.abort(&err);
		}

		Ok(())
	}

	async fn run_broadcast(self, path: PathOwned, mut broadcast: BroadcastProducer) {
		// Actually start serving subscriptions.
		loop {
			// Keep serving requests until there are no more consumers.
			// This way we'll clean up the task when the broadcast is no longer needed.
			let track = tokio::select! {
				_ = broadcast.unused() => break,
				producer = broadcast.requested_track() => match producer {
					Some(producer) => producer,
					None => break,
				},
				_ = self.session.closed() => break,
			};

			let id = self.next_id.fetch_add(1, atomic::Ordering::Relaxed);
			let mut this = self.clone();

			let path = path.clone();
			web_async::spawn(async move {
				this.run_subscribe(id, path, track).await;
				this.subscribes.lock().remove(&id);
			});
		}
	}

	async fn run_subscribe(&mut self, id: u64, broadcast: Path<'_>, track: TrackProducer) {
		self.subscribes.lock().insert(id, track.clone());

		self.control
			.send(
				ietf::MessageId::Subscribe,
				ietf::Subscribe {
					subscribe_id: id,
					track_alias: 0,
					track_namespace: broadcast.to_owned(),
					track_name: (&track.info.name).into(),
					subscriber_priority: track.info.priority,
				},
			)
			.ok();

		tracing::info!(id, broadcast = %self.origin.as_ref().unwrap().absolute(&broadcast), track = %track.info.name, "subscribe started");

		track.unused().await;
		tracing::info!(id, broadcast = %self.origin.as_ref().unwrap().absolute(&broadcast), track = %track.info.name, "subscribe cancelled");

		track.abort(Error::Cancel);
	}

	pub async fn recv_group(&mut self, stream: &mut Reader<S::RecvStream>) -> Result<(), Error> {
		let group: ietf::Group = stream.decode().await?;

		let group = {
			let mut subs = self.subscribes.lock();
			let track = subs.get_mut(&group.subscribe_id).ok_or(Error::Cancel)?;

			let group = Group {
				sequence: group.group_id,
			};
			track.create_group(group).ok_or(Error::Old)?
		};

		let res = tokio::select! {
			_ = group.unused() => Err(Error::Cancel),
			res = self.run_group(stream, group.clone()) => res,
		};

		match res {
			Err(Error::Cancel) | Err(Error::Transport(_)) => {
				tracing::trace!(group = %group.info.sequence, "group cancelled");
				group.abort(Error::Cancel);
			}
			Err(err) => {
				tracing::debug!(%err, group = %group.info.sequence, "group error");
				group.abort(err);
			}
			_ => {
				tracing::trace!(group = %group.info.sequence, "group complete");
				group.close();
			}
		}

		Ok(())
	}

	async fn run_group(&mut self, stream: &mut Reader<S::RecvStream>, mut group: GroupProducer) -> Result<(), Error> {
		while let Some(size) = stream.decode_maybe::<u64>().await? {
			let frame = group.create_frame(Frame { size });

			let res = tokio::select! {
				_ = frame.unused() => Err(Error::Cancel),
				res = self.run_frame(stream, frame.clone()) => res,
			};

			if let Err(err) = res {
				frame.abort(err.clone());
				return Err(err);
			}
		}

		group.close();

		Ok(())
	}

	async fn run_frame(&mut self, stream: &mut Reader<S::RecvStream>, mut frame: FrameProducer) -> Result<(), Error> {
		let mut remain = frame.info.size;

		tracing::trace!(size = %frame.info.size, "reading frame");

		while remain > 0 {
			let chunk = stream.read(remain as usize).await?.ok_or(Error::WrongSize)?;
			remain = remain.checked_sub(chunk.len() as u64).ok_or(Error::WrongSize)?;
			frame.write_chunk(chunk);
		}

		tracing::trace!(size = %frame.info.size, "read frame");

		frame.close();

		Ok(())
	}
}
